src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker_Generators.cs

```diff
@@ -178,25 +178,24 @@ private partial class CompilationTracker : ICompilationTracker
                         documentIdentity,
                         sourceText,
                         ProjectState.ParseOptions!,
-                        ProjectState.LanguageServices);
-                    Contract.ThrowIfTrue(generatedDocument.GetTextChecksum() != contentIdentity.Checksum, "Checksums must match!");
+                        ProjectState.LanguageServices,
+                        // Server provided us the checksum, so we just pass that along.  Note: it is critical that we do
+                        // this as it may not be possible to reconstruct the same checksum the server produced due to
+                        // the lossy nature of source texts.  See comment on GetOriginalSourceTextChecksum for more detail.
+                        contentIdentity.OriginalSourceTextChecksum);
+                    Contract.ThrowIfTrue(generatedDocument.GetOriginalSourceTextChecksum() != contentIdentity.OriginalSourceTextChecksum, "Checksums must match!");
```

Comment:
note: the assert is kinda meek at this point (since we're literally just validating the checksum we passed into the constructor is what we get back out), but i don't mind having it since if it does ever fail it will catch something immediately.